src/api/class-template-tags-request.php

```diff
@@ -46,14 +44,16 @@ public function check_template_tags() {
 	 * Calls and returns the list of template tags, categorized like in the API
 	 * response.
 	 *
-	 * @throws Api_Response_Error If API is down.
 	 * @return array Array of allowed template tags.
 	 */
 	private function get_template_tags() {
 		$tags_response = wp_remote_get( self::TEMPLATE_API_URL );
 
 		if ( is_wp_error( $tags_response ) ) {
-			throw Api_Response_Error::message( $tags_response->get_error_message() );
+			$this->error = $tags_response->get_error_message();
+			add_action( 'admin_notices', [ $this, 'notice' ] );
```

Comment:
Not sure why did you wrap this action in a method that is hooked to another action.

What I do is just hook to `admin_notices` and there I check for transients - if the transient is not `false` I add the error message from the transient to the error array and delete the transient, then just output the error notice out.

Reply:
Generally speaking it's usually best to load and execute code only when it's needed, so that where the `load-toplevel_page_theme-sniffer` hook comes in - to run the transient check/update only on the theme-sniffer page where it's needed.  The same with `admin_notices` - there's no need for this to hook directly to that and run a check on the transient every page load.  

In https://github.com/WPTRT/theme-sniffer/issues/165 the user was encountering the error as soon as the plugin was activated because of the fatal.  We should try to minimize impact of those type of situations.  By executing only where (and importantly when) code is needed we reduce our footprint.  So taking this example - even _if_ we introduced a fatal in this particular situation - the user would only experience this scenario while specifically being on our theme-sniffer page, instead of the plugin rendering the whole wp-admin unusable, just our page is unusable.  

Performance wise - if we go directly to `admin_notices` -  the normal cache of data in the transient isn't a huge query or anything like that, but the call using `wp_remote_get` _can_ be expensive - something we don't need to happen outside of context of the actual sniff page itself.  Using the code that's there, we aren't caching the error for any period of time, so in the event the remote API is unreachable, or the data for whatever reason isn't correct coming back, _every_ page load is trying to make the remote fetch.  This can lag an admin area down pretty quickly, especially for calls with longer timeouts. We still want to hook into `admin_notices` to display the caught error, but there's not any reason we need this to be displayed site wide, just whenever we encounter the error.  It's sorta the same principal as when you enqueue your css/js - you only ever need to enqueue your scripts/styles on the pages where they are needed.  The alternative in this case would be we _can_ still hook into `admin_notices` directly, and to achieve the same result, we check $pagenow global to ensure it's on the page needed.

If you look at Query Monitor plugin in the hooks & actions section, you'll kinda see the behavior a little better, where `load-toplevel_page_theme-sniffer` is more or less the admin page specific `init`, and when there is no error, no hook is added by this plugin in `admin_notices`.  If we hook directly to `admin_notices` and force an error, then the hook is added, and only when on our page, instead of all pages.

Reply:
> Generally speaking it's usually best to load and execute code only when it's needed, so that's where the load-toplevel_page_theme-sniffer hook comes in - to run the transient check/update only on the theme-sniffer page where it's needed. The same with admin_notices - there's no need for this to hook directly to that and run a check on the transient every page load.

Yeah, I usually have a return early check in the callback, if I'm not on a specific screen - bail out 🙂

Thanks for the detailed explanation, seems totally reasonable to add it this way, I'll refactor my code so that it works like this 😄